src/qutip_jax/linalg.py

```diff
@@ -12,23 +12,48 @@
     "eigs_jaxarray", "svd_jaxarray", "solve_jaxarray",
 ]
 
+def herm_with_vecs(data):
+    evals, evecs = jnp.linalg.eigh(data)
+    evals, evecs = evals.astype(jnp.complex64), evecs.astype(jnp.complex64)
+    return (evals, evecs)
 
-@partial(jit, static_argnums=[1, 2, 3, 4])
+def nonherm_with_vecs(data):
+    evals, evecs = jnp.linalg.eig(data)
+    evals, evecs = evals.astype(jnp.complex64), evecs.astype(jnp.complex64)
+    return (evals, evecs)
+
+def herm_no_vecs(data):
+    evals = jnp.linalg.eigvalsh(data)
+    evals = evals.astype(jnp.complex64)
+    dummy_evecs = jnp.zeros((data.shape[0], data.shape[0]), dtype=jnp.complex64)
+    return (evals, dummy_evecs)
+
+def nonherm_no_vecs(data):
+    evals = jnp.linalg.eigvals(data)
+    evals = evals.astype(jnp.complex64)
+    dummy_evecs = jnp.zeros((data.shape[0], data.shape[0]), dtype=jnp.complex64)
+    return (evals, dummy_evecs)
+
+
+@partial(jit, static_argnums=[2, 3, 4])
 def _eigs_jaxarray(data, isherm, vecs, eigvals, low_first):
     """
     Internal function to dispatch the eigenvalue solver to `eigh`, `eig`,
     `eigvalsh` or `eigvals` based on the parameters.
     """
-    if isherm and vecs:
-        evals, evecs = jnp.linalg.eigh(data)
-    elif vecs:
-        evals, evecs = jnp.linalg.eig(data)
-    elif isherm:
-        evals = jnp.linalg.eigvalsh(data)
-        evecs = None
-    else:
-        evals = jnp.linalg.eigvals(data)
-        evecs = None
+    def compute_with_vecs():
+        return lax.cond(
+            isherm, herm_with_vecs,
+            nonherm_with_vecs, data
+        )
+
+    def compute_no_vecs():
+        return lax.cond(
+            isherm, herm_no_vecs,
+            nonherm_no_vecs, data
+        )
+
+    evals, evecs = lax.cond(vecs, compute_with_vecs, compute_no_vecs)
```

Comment:
Since `vecs` is static, do we need to use `cond` here too?
Not having it would save us from creating dummy evecs that are not returned.

Reply:
Yes, you are right we do not. I will make the changes.

Reply:
@Ericgig Even if I remove `cond` here, we will still need four separate functions to compute the `evals` and `evecs`. Because it will still depend on `isherm`. So making changes won't save us from creating dummy evecs